clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java

```diff
@@ -620,6 +621,21 @@ protected Collection<PubSubMessage<K, ChangeEvent<V>, VeniceChangeCoordinate>> i
     if (changeCaptureStats != null) {
       changeCaptureStats.recordRecordsConsumed(pubSubMessages.size());
     }
+    if (changelogClientConfig.shouldCompactMessages()) {
+      Map<K, PubSubMessage<K, ChangeEvent<V>, VeniceChangeCoordinate>> tempMap = new LinkedHashMap<>();
+      // The behavior of LinkedHashMap is such that it maintains the order of insertion, but for values which are
+      // replaced,
+      // it's put in at the position of the first insertion. This isn't quite what we want, we want to keep only
+      // a single key (just as a map would), but we want to keep the position of the last insertion as well. So in order
+      // to do that, we remove the entry before inserting it.
+      for (PubSubMessage<K, ChangeEvent<V>, VeniceChangeCoordinate> message: pubSubMessages) {
```

Comment:
IIUC `pubSubMessages` could contain control messages if `includeControlMessage == true`. Could this break anything for CC client? I know heartbeat keys are equal but compacting heartbeat should is fine. What about other control messages?

Reply:
I wasn't able to think of a bad scenario.  Fortunately, the include control messages feature isn't really getting used by any customers yet, it's more used for internal interfaces right now.  I think we're good here for now.